### PR TITLE
Ostrich tlc tests

### PR DESCRIPTION
**Pull request by frost-ostrich**
# Description
What does this PR solve?

Fixes # (issue)

# Checklist
- [ ] Code follows the style guide
- [ ] I have performed a self-review of my own code
- [ ] New and existing tests pass locally
- [ ] This PR is targeting the correct branch

### CODE REVIEW

File: protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    assertEq(kinkInterestRate, testKinkInterestRate);
+    assertEq(kinkUtilizationRatio, testKinkUtilizationRatio);
+  }
+
+  function test_computeInterestRateImpl() public {
+    uint256 utilizationRation = mock.computeInterestRateImpl(expectedUtilizationRation);
+
+    assertEq(utilizationRation, expectedDaiInterestRate);
+
+  }
```
Comment (frost-ostrich) [inline protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:164]: @frontier159 this is throwing error here, I think I don't need to test `computeInterestRateImpl`? I just realized it is an internal function 😅

```
Compiler run failed:
Error (9582): Member "computeInterestRateImpl" not found or not visible after argument-dependent lookup in contract Mock.
   --> test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:160:33:
    |
160 |     uint256 utilizationRation = mock.computeInterestRateImpl(expectedUtilizationRation);
```

Comment (frontier159) [reply protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:164]: The visibility is just like in OO classes. Private == only visible to that contract, internal == visible to that contract and any contact which inherits it

File: protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import {ud} from '@prb/math/src/UD60x18.sol';
+
+import {TempleTest} from '../../TempleTest.sol';
+import {LinearWithKinkInterestRateModel} from 'contracts/v2/interestRate/LinearWithKinkInterestRateModel.sol';
+import {CommonEventsAndErrors} from 'contracts/common/CommonEventsAndErrors.sol';
+import {IInterestRateModel} from 'contracts/interfaces/v2/interestRate/IInterestRateModel.sol';
+import {FakeERC20} from 'contracts/fakes/FakeERC20.sol';
+import { SafeCast } from "contracts/common/SafeCast.sol";
+
+contract Mock is LinearWithKinkInterestRateModel {
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:12]: Hmm why do you need a mock contract? You can just create a LinearWithKink as the thing to test below

Comment (frost-ostrich) [reply protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:12]: If I want to test the `setRateParams` function and their access, I either need to use a mock contract, or add the function to the interface `IInterestRateModel`, which currently only have the `calculateInterestRate`, so that would be the only function I could test afaik, unless there is another way I'm not aware? 

I'll give it a go for now by adding `setRateParams` to `IInterestRateModel`

File: protocol/test/forge/v2/interestRate/CompoundedInterest.t.sol [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  uint256 public initialPrincipalAmount = 100e18;
+  
+}
+
+contract CompoundedInterestTestModifiers is CompoundedInterestTestBase {
+
+  // One percent interest tests
+  function test_computeInterestRateOnePctOneDayImpl() public {    
+    uint256 newInterestRate = CompoundedInterest.continuouslyCompounded(initialPrincipalAmount, 1 days, uint96(ONE_PCT_INTEREST));
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/CompoundedInterest.t.sol:37]: Just checking if you understand what:
`using CompoundedInterest for uint256;`
does?

It's syntactic sugar only for libraries only, but acts like the first param is a class/contract when plumbed through to the lib
In this case, if you have the `using ...`, then you can then do:

`uint256 newInterestRate = initialPrincipalAmount.continuouslyCompounded(1 days, uint96(ONE_PCT_INTEREST));`

File: protocol/test/forge/v2/interestRate/CompoundedInterest.t.sol [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import { TempleTest } from '../../TempleTest.sol';
+import { CompoundedInterest } from 'contracts/v2/interestRate/CompoundedInterest.sol';
+import { CommonEventsAndErrors } from 'contracts/common/CommonEventsAndErrors.sol';
+
+import { stdError } from "forge-std/Test.sol";
+import { PRBMath_MulDiv18_Overflow } from "@prb/math/src/Common.sol";
+
+contract CompoundedInterestTestBase is TempleTest {
+  using CompoundedInterest for uint256;
+  uint256 public constant ONE_PCT_INTEREST = 0.01e18;
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/CompoundedInterest.t.sol:13]: You could define this as a `uint96` to save having to cast it below (same as for the FIVE_PCT and TEN_PCT)

File: protocol/test/forge/v2/interestRate/CompoundedInterest.t.sol [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  function test_computeInterestRateTenPctThirtyDayImpl() public {    
+    uint256 newInterestRate = CompoundedInterest.continuouslyCompounded(initialPrincipalAmount, 30 days, uint96(TEN_PCT_INTEREST));
+    assertEq(newInterestRate, TEN_PCT_30DAY);
+  }
+  
+  function test_computeInterestRateTenPctOneYearImpl() public {    
+    uint256 newInterestRate = CompoundedInterest.continuouslyCompounded(initialPrincipalAmount, 365 days, uint96(TEN_PCT_INTEREST));
+    assertEq(newInterestRate, TEN_PCT_1YEAR);
+  }
+
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/CompoundedInterest.t.sol:82]: Can you make sure we have the other edge cases listed? What about 0 days, 0 principal, 0 interest rate?

Any others you can think of?

File: protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+import {ud} from '@prb/math/src/UD60x18.sol';
+
+import {TempleTest} from '../../TempleTest.sol';
+import {LinearWithKinkInterestRateModel} from 'contracts/v2/interestRate/LinearWithKinkInterestRateModel.sol';
+import {CommonEventsAndErrors} from 'contracts/common/CommonEventsAndErrors.sol';
+import {IInterestRateModel} from 'contracts/interfaces/v2/interestRate/IInterestRateModel.sol';
+import { SafeCast } from "contracts/common/SafeCast.sol";
+
+contract LinearWithKinkInterestRateModelTestBase is TempleTest {
+  LinearWithKinkInterestRateModel public daiInterestRateModel;
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:12]: Please be consistent with the other solidity for spacing - should be 4 spaces

File: protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      365 days
+    ); // ~54k
+
+    expectedUtilizationRatio = utilizationRatio(
+      expectedDaiDebt,
+      BORROW_CEILING
+    ); // ~54%
+  }
+
+  function utilizationRatio(
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:50]: Do you think having the UR calc embedded in this unit test makes it a bit more complicated than it needs to be?

Perhaps just hand-picking the actual UR to use as input to the function would be simpler than deriving it?

File: protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+{
+  event InterestRateParamsSet(
+    uint256 _baseInterestRate,
+    uint256 _maxInterestRate,
+    uint256 _kinkUtilizationRatio,
+    uint256 _kinkInterestRate
+  );
+
+  using SafeCast for uint256;
+
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:89]: I find it's good to add a test after construction to ensure it's doing the expected thing there too (eg in case i forgot to set some parameter)

File: protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+      vm.expectRevert(
+        abi.encodeWithSelector(CommonEventsAndErrors.InvalidAccess.selector)
+      );
+      daiInterestRateModel.setRateParams(0.02e18, 0.75e18, 0.5e18, 0.1e18);
+      vm.stopPrank();
+    }
+
+    // access for executor
+    {
+      vm.startPrank(executor);
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:103]: This is just PSA -- there's another cheatcode you can use `changePrank(executor)` which does the `vm.stopPrank(); vm.startPrank(executor);` in one line

File: protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+  function test_expectedInterestRate() public {
+    assertEq(expectedDaiInterestRate, 0.077777777777777777e18); // ~7%
+  }
+  function test_expectedDebtAmount() public {
+    assertEq(expectedDaiDebt, 54_044.121788100384500000e18); // ~54k
+  }
+  function test_expectedUtilizationRatio() public {
+    assertEq(expectedUtilizationRatio, 0.540441217881003845e18); // ~54%
+  }
+}
```
[the remark this answers is not in the document]
Comment (frost-ostrich) [reply protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:295]: - Coverage 100% for `LinearWithKinkInterestRateModel.sol`
- Coverage still not 100 for `BaseInterestRateModel.sol`, it may be because it's an abstract contract
- Coverage not working for `CompoundedInterest.sol`, prob because is a library?

image

File: .prettierrc [hunk 11]
```diff
@@ -3,5 +3,14 @@
...
   "singleQuote": true,
-  "useTabs": false
-}
\ No newline at end of file
+  "overrides": [
+    {
+      "files": "*.sol",
+      "options": {
+        "semi": true,
+        "singleQuote": true,
```
Comment (frontier159) [inline .prettierrc:11]: Need double quotes for solidity

File: protocol/test/forge/v2/interestRate/CompoundedInterest.t.sol [hunk 12]
```diff
@@ -0,0 +1,190 @@
+pragma solidity ^0.8.17;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+import {TempleTest} from '../../TempleTest.sol';
+import {CompoundedInterest} from 'contracts/v2/interestRate/CompoundedInterest.sol';
+import {CommonEventsAndErrors} from 'contracts/common/CommonEventsAndErrors.sol';
+
+import {stdError} from 'forge-std/Test.sol';
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/CompoundedInterest.t.sol:8]: Both `CommonEventsAndErrors` and `stdError` are unused - could you cleanup?

File: protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol [hunk 13]
```diff
@@ -0,0 +1,306 @@
+pragma solidity ^0.8.17;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+import {ud} from '@prb/math/src/UD60x18.sol';
+
+import {TempleTest} from '../../TempleTest.sol';
+import {LinearWithKinkInterestRateModel} from 'contracts/v2/interestRate/LinearWithKinkInterestRateModel.sol';
+import {CommonEventsAndErrors} from 'contracts/common/CommonEventsAndErrors.sol';
+import {IInterestRateModel} from 'contracts/interfaces/v2/interestRate/IInterestRateModel.sol';
```
Comment (frontier159) [inline protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:8]: Both `IInterestRateModel` and `ud` are unused - could you cleanup?

**Comment by frost-ostrich** [comment]
closing this in favour of https://github.com/TempleDAO/temple/pull/816
